### PR TITLE
Add KKTIX attendee Info ETL to BigQuery pycontw-225217.dwd.kktix_ticket_xxxx_attendees tables

### PR DESCRIPTION
**Pull request by GeekAngus**
## Types of changes

- **New feature**

## Description
Background: Start from 2022, we extract the KKTIX data via KKTIX API and load to "pycontw-225217.ods.ods_kktix_attendeeId_datetime". However most of the data are store in the ATTENDEE_INFO column with json format. To use metabase with SQL, users need to extract the data by json_extract with the knowledge kktix format instead of flat database. And we also need to rewrite all the SQLs build for current databases.
Solution: Transform the tables in backend that we could keep the same user experience by using Metabase.

## Checklist:
- [ ] Add test cases to all the changes you introduce
- [X] Run `poetry run pytest` locally to ensure all linter checks pass
- [X] Update the documentation if necessary

## Steps to Test This Pull Request
`cd contrib`
`./kktix_bq_etl.sh 2023`

## Expected behavior
The data had been load to dwd.kktix_ticket_${ticket_type}_attendees_test on bigquery
ticket_type = corporate individual reserved

### CODE REVIEW

File: contrib/README.md [hunk 1]
```diff
@@ -17,4 +17,10 @@
...
     2. If everything looks good, you can `copy` the `fact table` and `dimension table` first
-    3. Then run `python upload-survey-cake-csv-to-bigquery.py --year=<20xx> -p`. `-p` stands for `production`
\ No newline at end of file
+    3. Then run `python upload-survey-cake-csv-to-bigquery.py --year=<20xx> -p`. `-p` stands for `production`
+
+## KKTIX BigQuery Transform
+1. Background: Start from 2022, we extract the KKTIX data via KKTIX API and load to "pycontw-225217.ods.ods_kktix_attendeeId_datetime". However most of the data are store in the ATTENDEE_INFO column with json format. To use metabase with SQL, users need to extract the data by json_extract with the knowledge kktix format instead of flat database. And we also need to rewrite all the SQLs build for current databases.
+2. Solution: Transform the tables in backend that we could keep the same user experience by using Metabase.
+3. Run: 
+ - for 3 tables in single bash script: `./kktix_bq_etl.sh 2023`
```
Comment (david30907d) [inline contrib/README.md:26]: thx for documenting these up 🙏

File: dags/ods/kktix_ticket_orders/udfs/kktix_loader.py [hunk 2]
```diff
@@ -28,6 +30,7 @@ def load(event_raw_data_array: List):
         sanitized_event_raw_data = _sanitize_payload(event_raw_data)
         payload.append(sanitized_event_raw_data)
     _load_to_bigquery(payload)
+    _load_to_bigquery_dwd(payload)
```
Comment (david30907d) [inline dags/ods/kktix_ticket_orders/udfs/kktix_loader.py:33]: 👍

File: dags/ods/kktix_ticket_orders/udfs/kktix_bq_dwd_etl.py [hunk 3]
```diff
@@ -0,0 +1,667 @@
...
+    # print(res_clone)
+
+    print(f"Extracted from {TABLE}  successful.")
+
+    # load to DataFrame for later bigquery upload from the extracted results
+    df, sanitized_df = load_to_df_from_list(results, "bigquery", update_after_ts)
+    # print(sanitized_df.columns)
+    # print(sanitized_df.head())
+    # df_null = sanitized_df.isnull()
+    # print(sanitized_df.iloc[:, :5])
```
Comment (david30907d) [inline dags/ods/kktix_ticket_orders/udfs/kktix_bq_dwd_etl.py:646]: should we remove these comments?

File: dags/ods/kktix_ticket_orders/udfs/kktix_bq_dwd_etl.py [hunk 4]
```diff
@@ -0,0 +1,667 @@
...
+    df_data[column_list] = [value_list]
+
+    df_dict = pd.concat([df_dict, df_data], axis=1)
+    return df_dict
+
+
+def load_to_df_from_list(
+    results, source="dag", update_after_ts=0
+) -> Tuple[DataFrame, DataFrame]:
+    # Use DataFrame for table transform operations
```
Comment (david30907d) [inline dags/ods/kktix_ticket_orders/udfs/kktix_bq_dwd_etl.py:493]: seems to me that you should rename this function to `transform_xxx()`?

File: dags/ods/kktix_ticket_orders/udfs/kktix_bq_dwd_etl.py [hunk 5]
```diff
@@ -0,0 +1,667 @@
...
+SCHEMA = [
+    bigquery.SchemaField("id", "INTEGER", mode="REQUIRED"),
+    bigquery.SchemaField("name", "STRING", mode="REQUIRED"),
+    bigquery.SchemaField("attendee_info", "STRING", mode="REQUIRED"),
+    bigquery.SchemaField("refunded", "BOOLEAN", mode="REQUIRED"),
+]
+JOB_CONFIG = bigquery.LoadJobConfig(schema=SCHEMA)
+
+
+def _load_row_df_from_dict(json_dict, update_after_ts) -> DataFrame:
```
Comment (david30907d) [inline dags/ods/kktix_ticket_orders/udfs/kktix_bq_dwd_etl.py:429]: ditto

File: dags/ods/kktix_ticket_orders/udfs/kktix_bq_dwd_etl.py [hunk 6]
```diff
@@ -0,0 +1,667 @@
...
+    "email_to_sponsor",
+    "ive_already_read_and_i_accept_the_epidemic_prevention_of_pycon_tw",
+    "ive_already_read_and_i_accept_the_privacy_policy_of_pycon_tw",
+    "email",
+    "registration_no",
+    "attendance_book",
+}
+
+CANONICAL_COLUMN_NAMES_2020_EXTRA_CORPORATE = {
+    "invoice_policy",
```
Comment (david30907d) [inline dags/ods/kktix_ticket_orders/udfs/kktix_bq_dwd_etl.py:56]: unused constants?

File: contrib/kktix_bq_etl.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+cmd=${PWD}/../dags/ods/kktix_ticket_orders/udfs/kktix_bq_dwd_etl.py
+
+
+for ticket_type in corporate individual reserved
+do
+    suffix=${ticket_type}_attendees$2
+    cmd_args="-p ${project_id} -d dwd -t kktix_ticket_${suffix} -k ${ticket_type} -y $1 --upload"
+    echo ${cmd_args}
+    ${cmd} ${cmd_args}
+done
```
Comment (david30907d) [inline contrib/kktix_bq_etl.sh:15]: 👍

File: dags/ods/kktix_ticket_orders/udfs/kktix_loader.py [hunk 8]
```diff
@@ -43,6 +46,30 @@ def _load_to_bigquery(payload: List[Dict]) -> None:
     job.result()
 
 
+def _load_to_bigquery_dwd(payload: List[Dict]) -> None:
```
Comment (david30907d) [inline dags/ods/kktix_ticket_orders/udfs/kktix_loader.py:49]: sorry for the nit picking, but seems to me that this function is more like `transform` than `load`

**Review by david30907d — APPROVED** [review]
Generally, it looks good to me. I just have some suggestions regarding the design aspect.

So feel free to ship it, and we can discuss the design pattern afterwards

## About the Design Pattern

Hera are my 2 cents:
1. Follow `ETL` or `ELT` paradigm (for example: https://github.com/pycontw/pycon-etl/blob/master/dags/ods/kktix_ticket_orders/udfs/kktix_api.py#L30-L34)
2. Follow single responsibility principle: Since `_load_to_bigquery_dwd()` is actually doing transformation, (https://github.com/pycontw/pycon-etl/blob/cd68ec1cca12df5ca1c95767cc52963025ab2076/dags/ods/kktix_ticket_orders/udfs/kktix_loader.py#L20-L33) is no longer following `single responsibility principle`. I would suggest using an independent operator for this `dwd` operation.
3. Make the pipeline idempotent if possible. Since you're using `WriteDisposition.WRITE_APPEND` right now, it means we might have duplicate data if someone re-runs a partially failed job.

**Review by henry410213028 — APPROVED** [review]

**Pull request merged**